src/librustc_infer/infer/error_reporting/nice_region_error/static_impl_trait.rs

```diff
@@ -2,229 +2,300 @@
+        err.span_label(param.param_ty_span, &format!("this data with {}...", lifetime));
+        debug!("try_report_static_impl_trait: param_info={:?}", param);
+
+        // We try to make the output have fewer overlapping spans if possible.
+        if (sp == sup_origin.span() || !return_sp.overlaps(sup_origin.span()))
+            && sup_origin.span() != return_sp
+        {
+            // FIXME: account for `async fn` like in `async-await/issues/issue-62097.rs`
+
+            // Customize the spans and labels depending on their relative order so
+            // that split sentences flow correctly.
+            if sup_origin.span().overlaps(return_sp) && sp == sup_origin.span() {
+                // Avoid the following:
+                //
+                // error: cannot infer an appropriate lifetime
+                //   --> $DIR/must_outlive_least_region_or_bound.rs:18:50
+                //    |
+                // LL | fn foo(x: &i32) -> Box<dyn Debug> { Box::new(x) }
+                //    |           ----                      ---------^-
+                //
+                // and instead show:
+                //
+                // error: cannot infer an appropriate lifetime
+                //   --> $DIR/must_outlive_least_region_or_bound.rs:18:50
+                //    |
+                // LL | fn foo(x: &i32) -> Box<dyn Debug> { Box::new(x) }
+                //    |           ----                               ^
+                err.span_label(
+                    sup_origin.span(),
+                    "...is captured here, requiring it to live as long as `'static`",
                 );
-                let anon_reg_sup = self.tcx().is_suitable_region(sup_r)?;
-                debug!("try_report_static_impl_trait: anon_reg_sup={:?}", anon_reg_sup);
-                let fn_returns = self.tcx().return_type_impl_or_dyn_traits(anon_reg_sup.def_id);
-                if fn_returns.is_empty() {
-                    return None;
-                }
-                debug!("try_report_static_impl_trait: fn_return={:?}", fn_returns);
-                if **sub_r == RegionKind::ReStatic {
-                    let sp = var_origin.span();
-                    let return_sp = sub_origin.span();
-                    let param_info = self.find_param_with_region(sup_r, sub_r)?;
-                    let (lifetime_name, lifetime) = if sup_r.has_name() {
-                        (sup_r.to_string(), format!("lifetime `{}`", sup_r))
-                    } else {
-                        ("'_".to_owned(), "an anonymous lifetime `'_`".to_string())
-                    };
-                    let mut err = struct_span_err!(
-                        self.tcx().sess,
-                        sp,
-                        E0759,
-                        "cannot infer an appropriate lifetime"
+            } else {
+                err.span_label(sup_origin.span(), "...is captured here...");
+                if return_sp < sup_origin.span() {
```

Comment:
It would be nice for `DiagnosticBuilder` to have a concept of 'ordered diagnostics', as there are lots of other places where this kind of logic would be useful. However, it doesn't have to block this PR.

Reply:
I had a proposal around this at some point. I wanted to change the rendering if the diagnostics would not display in source order.